### DISCUSSION TITLE
Animation resume/pause support

### DISCUSSION THREAD

**Discussion by chonwhite** [slint-ui/slint#3920] [Q&A] [answered]
I'm using Image rotation animation to emulate a vinyl spin, and need to pause and resume the animation based on some state change. Can't find any documentation mentioning animation pause and resume. My main language is C++, BTW.

**Comment by ogoffart** [comment, marked as answer]
Thanks for your issue. (I moved it into discussion to keep the issues for bug tracking)

The way to do permanently running animation is to use the animation-tick() builtin function.
Depending on the state, you can use it to make the angle vary, or not.
[Example](https://slintpad.com/?snippet=import+%7B+AboutSlint%2C+Button%2C+VerticalBox+%7D+from+%22std-widgets.slint%22%3B%0Aexport+component+Demo+%7B%0A%0A++++property+%3Cbool%3E+running%3A+true%3B%0A++++property+%3Cduration%3E+previous-press%3B%0A%0A++++property+%3Cangle%3E+base_angle%3B%0A%0A++++VerticalBox+%7B%0A++++++++alignment%3A+start%3B%0A++++++++img+%3A%3D+Image+%7B%0A++++++++++++source%3A+%40image-url%28%22https%3A%2F%2Fslint.dev%2Flogo%2Fslint-logo-full-light.svg%22%29%3B%0A++++++++++++height%3A+100px%3B%0A++++++++++++rotation-angle%3A+running+%3F+base-angle+%2B+%28animation-tick%28%29+-+previous-press%29+*+1turn+%2F+3s+%3A+base-angle%3B%0A++++++++%7D%0A++++++++HorizontalLayout+%7B+%0A++++++++++++alignment%3A+center%3B%0A++++++++++++Button+%7B+%0A++++++++++++++++text%3A+%22Play%2FPause%22%3B+%0A++++++++++++++++clicked+%3D%3E+%7B+base-angle+%3D+img.rotation-angle%3B+previous-press+%3D+animation-tick%28%29%3B++running+%3D+%21running+%7D+%0A++++++++++++%7D+%0A++++++++%7D%0A++++%7D%0A%7D%0A&style=native)

**Comment by chonwhite** [comment]
Thanks, works like a charm!

**Comment by chonwhite** [comment]
A small issue though, when the animation is applied, it consumes quite a bit of CPU, about 15% on my M2 MacBook Pro， maybe could be optimised later?

**Reply by ogoffart** [reply]
Are you building in release mode?
And what renderer are you using? the default is femtovg, but you can try the skia renderer, it might be faster

**Reply by chonwhite** [reply]
Building under Release mode, the whole project was based on [slint-cpp-template](https://github.com/slint-ui/slint-cpp-template), and no changes were made to project settings.

**Comment by chonwhite** [comment]
Replicated the same animation with Objective-C on MacOSX and here's the performance result.
<img width="258" alt="image" src="https://github.com/slint-ui/slint/assets/1076429/e6a82bea-ed28-4539-9d15-41b1e70b44bc">

**Reply by tronical** [reply]
I think we're missing frame throttling on macOS (as well as others). The issue is particularly visible on macOS - it also causes delays with accessibility updates. Winit now has [support](https://docs.rs/winit/latest/winit/window/struct.Window.html#method.pre_present_notify) for this on Wayland. I've filed https://github.com/slint-ui/slint/issues/4025 for our side.
